crucible-llvm/src/Lang/Crucible/LLVM/Translation/BlockInfo.hs

```diff
@@ -0,0 +1,280 @@
+
+
+-- | Compute predecessor sets from the successor sets already computed in @buildBlockInfo@
+updatePredSets :: LLVMBlockInfoMap s -> LLVMBlockInfoMap s
+updatePredSets bim0 = foldl' upd bim0 predEdges
+  where
+   upd bim (to,from) = Map.adjust (\bi -> bi{ block_pred_set = Set.insert from (block_pred_set bi) }) to bim
+
+   predEdges =
+     [ (to,from) | (from, bi) <- Map.assocs bim0
+                 , to <- Set.elems (block_succ_set bi)
+     ]
+
+-- | Compute use sets for each basic block. Thess sets list all the
+-- virtual registers that must be assigned before jumping to a
+-- block.
+--
+-- This is essentially a backward fixpoint computation based on the
+-- identifiers used in the block statements.  We iterate the use/def
+-- transfer function until no more changes are made.  Because it is a
+-- backward analysis, we (heuristicly) examine the blocks in
+-- descending order, and re-add blocks to the workset based on
+-- predecessor maps.
+--
+-- This fixpoint computation terminates for the usual reasons: the transfer
+-- function is monotonic (use sets only increase) and has no infinite
+-- chains (in the worst case, all the finitely-many identifers in the
+-- function end up in every use set).
+computeUseSets :: LLVMBlockInfoMap s -> LLVMBlockInfoMap s
+computeUseSets bim0 = loop bim0 (Map.keysSet bim0) -- start with all blocks in the workset
+  where
+  loop bim ws =
+    -- chose the largest remaining block label in the workset
+    case Set.maxView ws of
+      -- if the workset is empty, we are done
+      Nothing -> bim
+      Just (l, ws') ->
+        -- look up the current block information relating to block l
+        case Map.lookup l bim of
+          Nothing -> panic "computeUseSets" ["Could not find label", show l]
+          Just bi ->
+            -- run the transfer function and compute an updated use set
+            case updateUseSet l bi bim of
+              -- if there is no update, continue down the work set
+              Nothing -> loop bim ws'
+              -- if we updated the use set, record it in the block map and
+              -- add all the predecessors of this block back to the work set
+              Just bi' ->
+                loop (Map.insert l bi' bim) (Set.union ws' (block_pred_set bi'))
+
+-- | Run the use/def transfer function on the block body and update the block info if
+-- any changes are detected
+updateUseSet :: L.BlockLabel -> LLVMBlockInfo s -> Map L.BlockLabel (LLVMBlockInfo s) -> Maybe (LLVMBlockInfo s)
+updateUseSet lab bi bim = if newuse == block_use_set bi then Nothing else Just bi{ block_use_set = newuse }
+  where
+  newuse = loop (block_body bi)
+
+  loop [] = mempty
+  loop (L.Result nm i _md:ss) = Set.union (instrUse lab i bim) (Set.delete nm (loop ss))
+  loop (L.Effect i _md:ss) = Set.union (instrUse lab i bim) (loop ss)
+
+instrUse :: L.BlockLabel -> L.Instr -> Map L.BlockLabel (LLVMBlockInfo s) -> Set L.Ident
+instrUse from i bim = Set.unions $ case i of
```

Comment:
Curiously, this function only calls `Set.unions` once at the top level, whereas `useVal` invokes `Set.unions` recursively. Is there a good reason not to pick a single convention for both functions?